File: chart.py
class BarChart:
    """A bar chart.

    This is a base chart object that assumes data has been processed such that:

        1. x-coordinates are integers from 0-len(data),
        2. y-coordinates are integers from 0-height * nbars,

    where `nbars` is the number of

    Args:
        data: a list of values to be plotted
        height: the number of rows to use
    """

    def __init__(self, data: list[int], height: int):
        self.data = data
        self.height = height

    def render(self) -> str:
        chart = ""
        for row_index in range(self.height, -1, -1):
            chart += BarChartRow(row_index, self.data).render()

        return chart


class BarChartRow:
    bars = ["▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"]
    nbars = len(bars)

    def __init__(self, index: int, data: list[int]):
        self.index = index
        self.data = data

    @property
    def min_y(self):
        return self.index * self.nbars

    @property
    def max_y(self):
        return (self.index + 1) * self.nbars

    def render(self) -> str:
        row = ""
        for value in self.data:
            if value >= self.max_y:
                row += self.bars[self.nbars - 1]
            elif value > self.min_y:
                row += self.bars[value % self.nbars - 1]
            else:
                row += " "

        return row + "\n"


def xrange(start, stop, n):
    step = (stop - start) / n
    return [start + step * i for i in range(n + 1)]


class Histogram:
    """A textual histrogram based.

    The histogram represents *y-weighted* counts of x values. That is, the height of bars is equal to the sum of the `y[i]` for each `x[i]` in the bar's bin. To produce a standard histogram of x values, use `y = ones(len(x))`.

    - x-values are binned to a length-`width` grid [x_min - x_tol, ..., x_max + x_tol]
    - y-values are binned to a length-`n x height + 1` grid [y_min - y_tol, ..., y_max + y_tol],

    where `n` is the number of steps per row (default: 8). The binning process results in a new length-`width` data vector, `data`, where each element is in `[0, 1, ..., n x height]`.

    Args:
        x: list of x-values to plot
        y: list of y-values to plot
        width: the number of columns to use
        height: the number of rows to use
    """

    bars = ["▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"]
    nbars = len(bars)

    def __init__(
        self,
        x: list[float],
        y: list[int],
        width: int,
        height: int,
        x_min: float | None,
        x_max: float | None,
    ):
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.x_min = x_min
        self.x_max = x_max

    def data(self) -> list[int]:
        import math

        self.x_min = self.x_min if self.x_min is not None else min(self.x)
        self.x_max = self.x_max if self.x_max is not None else max(self.x)
        self.x_step = (self.x_max - self.x_min) / self.width

        data = [0 for _ in range(self.width)]
        for xindex, value in enumerate(self.x):
            dindex = int(math.floor((value - self.x_min) / self.x_step))
            if dindex == self.width:
                dindex -= 1
            data[dindex] += self.y[xindex]

        self.y_min = 0
        self.y_max = max(data)
        self.y_step = (self.y_max - self.y_min) / (self.height * self.nbars)

        data = [int(math.floor(value / self.y_step)) for value in data]

        return data

    def render(self) -> str:
        return BarChart(self.data(), self.height).render()


import random

x = [random.random() for _ in range(10_000)]
# y = [random.randint(1, 10) for _ in range(10_000)]
y = [1 for _ in range(10_000)]
print(Histogram(x, y, 20, 10, x_min=0.0, x_max=1.0).render())

File: source/components/Chart.js
import React from 'react';
import {Box, Text} from 'ink';

/**
 * A textual histogram component for Ink.js
 *
 * Creates a histogram representing y-weighted counts of x values.
 * The height of bars equals the sum of y[i] for each x[i] in the bar's bin.
 * For a standard histogram of x values, use y = ones(len(x)).
 *
 * - x-values are binned to a length-width grid [xMin - xTol, ..., xMax + xTol]
 * - y-values are binned to a length-(nbars * height + 1) grid [yMin - yTol, ..., yMax + yTol]
 *
 * @param {Object} props
 * @param {number[]} props.x - Array of x-values to plot
 * @param {number[]} props.y - Array of y-values to plot (weights for each x)
 * @param {number} props.width - Number of columns/bins to use
 * @param {number} props.height - Number of rows to use
 * @param {number} [props.xMin] - Minimum x value (defaults to min of x)
 * @param {number} [props.xMax] - Maximum x value (defaults to max of x)
 * @param {number} [props.yMin] - Minimum y value for scaling (defaults to 0)
 * @param {number} [props.yMax] - Maximum y value for scaling (defaults to max of binned data)
 * @param {string} [props.color] - Optional color for the bars
 */
export function Histogram({
	x,
	y,
	width,
	height,
	xMin,
	xMax,
	yMin,
	yMax,
	color,
}) {
	const bars = ['▁', '▂', '▃', '▄', '▅', '▆', '▇', '█'];
	const nbars = bars.length;

	// Calculate binned data
	const calculateData = () => {
		const actualXMin = xMin !== undefined ? xMin : Math.min(...x);
		const actualXMax = xMax !== undefined ? xMax : Math.max(...x);
		const xStep = (actualXMax - actualXMin) / width;

		// Initialize bins
		const data = Array(width).fill(0);

		// Bin x values and accumulate y values
		x.forEach((value, xIndex) => {
			let dIndex = Math.floor((value - actualXMin) / xStep);
			if (dIndex === width) {
				dIndex -= 1;
			}
			if (dIndex >= 0 && dIndex < width) {
				data[dIndex] += y[xIndex];
			}
		});

		// Calculate y scaling
		const actualYMin = yMin !== undefined ? yMin : 0;
		const actualYMax = yMax !== undefined ? yMax : Math.max(...data);
		const yStep = (actualYMax - actualYMin) / (height * nbars);

		// Scale data to chart height
		const scaledData = data.map(value =>
			Math.floor((value - actualYMin) / yStep),
		);

		return scaledData;
	};

	const chartData = calculateData();

	return <BarChart data={chartData} height={height} color={color} />;
}

/**
 * A textual bar chart component for Ink.js
 *
 * Renders data as a vertical bar chart using Unicode block characters.
 * Data values should be integers from 0 to height * nbars, where nbars is 8 (the number of block characters).
 *
 * @param {Object} props
 * @param {number[]} props.data - Array of integer values to plot
 * @param {number} props.height - Number of rows to use for the chart
 * @param {string} [props.color] - Optional color for the bars
 */
function BarChart({data, height, color}) {
	const bars = ['▁', '▂', '▃', '▄', '▅', '▆', '▇', '█'];
	const nbars = bars.length;

	const renderRow = (rowIndex, data) => {
		const minY = rowIndex * nbars;
		const maxY = (rowIndex + 1) * nbars;

		return data
			.map((value, idx) => {
				let char;
				if (value >= maxY) {
					char = bars[nbars - 1];
				} else if (value > minY) {
					char = bars[(value % nbars) - 1];
				} else if (rowIndex === 0) {
					char = bars[0];
				} else {
					char = ' ';
				}
				return char;
			})
			.join('');
	};

	const rows = [];
	for (let rowIndex = height; rowIndex >= 0; rowIndex--) {
		rows.push(
			<Text key={rowIndex} color={color}>
				{renderRow(rowIndex, data)}
			</Text>,
		);
	}

	return <Box flexDirection="column">{rows}</Box>;
}
